File: algo/199. Binary Tree Right Side View.js
// BFS

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function(root) {
    if (root == null) {
        return [];
    }
    
    res = [];
    var queue = [root];
    while (queue.length != 0) {
        var levelSize = queue.length;
        for (var i=0; i < levelSize; i++) {
            var currNode = queue.shift();
            if (i==levelSize-1) {
                res.push(currNode.val);
            }
            
            if (currNode.left) {
                queue.push(currNode.left);
            }
            
            if (currNode.right) {
                queue.push(currNode.right);
            }
        }        
    }
    
    return res;
};

File: round2/199. Binary Tree Right Side View.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<int> RightSideView(TreeNode root) {
        var queue = new Queue<TreeNode>();
        var res = new List<int>();
        if (root == null) return res;
        queue.Enqueue(root);
        while (queue.Count != 0) {
            int levelSize = queue.Count;
            for (int i=0; i<levelSize; i++) {
                var curr = queue.Dequeue();
                if (i == levelSize - 1) res.Add(curr.val);
                if (curr.left != null) queue.Enqueue(curr.left);
                if (curr.right != null) queue.Enqueue(curr.right);
            }
        }
        return res;
    }
}